### PR TITLE
Don't set null SecurityManager on Java 17

### PR DESCRIPTION
**Pull request by offa**
Sets a null Security Manager on Java < 17 only.

- #507 
- #481
- [JENKINS-67000](https://issues.jenkins.io/browse/JENKINS-67000)
- [JENKINS-67913](https://issues.jenkins.io/browse/JENKINS-67913)

--------------------------------

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -181,6 +181,11 @@ THE SOFTWARE.
       <version>9.3</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.jenkins-ci</groupId>
+      <artifactId>version-number</artifactId>
```
Comment (jglick) [inline pom.xml:186]: Probably fine but we will see when this is integrated into core whether there is any clash with the version used by core. There are also integration tests in core which `java -jar agent.jar` so any shading/packaging issues should be smoked out.

File: src/main/java/hudson/remoting/jnlp/Main.java [hunk 2]
```diff
@@ -246,12 +249,12 @@ public static void main(String[] args) throws IOException, InterruptedException
      * Main without the argument handling.
      */
     public static void _main(String[] args) throws IOException, InterruptedException, CmdLineException {
-        // TODO skip this on Java 17+ (e.g. io.jenkins.lib.versionnumber.JavaSpecificationVersion) as it prints a warning
-        // otherwise needed for JavaWebStart agents (JENKINS-67000)
```
Comment (jglick) [inline src/main/java/hudson/remoting/jnlp/Main.java:250]: I think you want to keep some comments explaining the context, specifically https://github.com/jenkinsci/remoting/pull/507#issuecomment-1030341067

**Review by jglick — APPROVED** [review]

**Review by basil — COMMENTED** [review]
My understanding is that avoiding the call to `System.setSecurityManager(null)` on Java 17+ will eliminate the warning but will also remove support for Java Web Start agents, as it did on Java 11 in [JENKINS-67000](https://issues.jenkins.io/browse/JENKINS-67000). In other words, since Java Web Start remains an option offered in the GUI on all Java versions (including Java 17+), this PR will just reintroduce JENKINS-67000 (but for Java 17 rather than Java 11).

Just because Java 17+ support is in preview does not mean we encourage the reintroduction of bugs. I think that in order to move forward with this PR we would either need to formally deprecate and then remove the Java Web Start feature when running on Java 17 (including removing the option in the GUI). Alternatively, we could find a way to implement the Web Start feature on Java 17 without the call to `System.setSecurityManager(null)`.

I have not investigated the latter yet, but certainly retaining support for this feature would not break any existing use cases. I am not particularly against the former, but there are at least some users for whom this is a production use case, so we would need to communicate with them in advance through a blog post and/or deprecation warning to ensure they are not surprised by a sudden change.

**Comment by timja** [comment]
JEP-236 says that we will be removing support as of Java 11.
https://github.com/jenkinsci/jep/tree/master/jep/236#library-updates

**Comment by basil** [comment]
> JEP-236 says that we will be removing support as of Java 11. https://github.com/jenkinsci/jep/tree/master/jep/236#library-updates

I had no part in the writing of JEP-236 and will not comment on it. My existing comments above remain. It is fine to deprecate and subsequently remove a feature so long as there is a cogent justification, this is communicated to users, and the removal happens at all levels (including GUI). In this case, deprecation has not yet been done, there is no communication plan, and the GUI remains, so it is premature to remove implementation support for this feature.

File: pom.xml (hunk 1 above)
Comment (offa) [reply pom.xml:186]: Core uses the same version (1.10) according to a quick GH search.

**Comment by offa** [comment]
Thanks @basil, I didn't have this on my radar. Since it's no longer shipped I expected no one to actually continue using it.

However, this complicates things … and I honestly have no idea how to continue here.

File: src/main/java/hudson/remoting/jnlp/Main.java (hunk 2 above)
Comment (offa) [reply src/main/java/hudson/remoting/jnlp/Main.java:250]: Indeed, added.

**Comment by jglick** [comment]
> Java Web Start remains an option offered in the GUI on all Java versions (including Java 17+)

As noted in https://github.com/jenkinsci/remoting/pull/507#issuecomment-1030341067 the version check is done on the controller, which is somewhat off base: if you run the controller on Java 8 you would still be offered the option to use JWS for agents, which would now break (I suppose) if your agent happened to use a 17 JRE and you had downloaded IcedTea or similar. A pretty obscure scenario but some users might be affected.

Options:
* Take the opportunity to formally remove JWS support for good. That means cleaning up some other stuff in `remoting` (GUI mode, JAR signature) and `jenkins` (link to launch via `javaws`), deprecating `*slave-installer-module` and removing these from the WAR, and properly announcing this as a breaking change with a migration guide.
* Somehow detect that the agent is being run from `javaws` (is there a system property you can introspect?) and retain the nulling out of the `SecurityManager` even on 17. For now this will only print a console warning—which is not visible from a JWS GUI launch anyway. Kicks the can down the road a bit, since a subsequent Java version will likely throw an error.

**Comment by MarkEWaite** [comment]
As a related aside, https://www.jenkins.io/doc/administration/requirements/upgrade-java-guidelines/#java-web-start states that Java Web Start is no longer available with Jenkins when the controller is running Java 11.  Not a blog post, but it has been publicly stated for a while

**Comment by basil** [comment]
bq. Take the opportunity to formally remove JWS support for good. That means cleaning up some other stuff in remoting (GUI mode, JAR signature) and jenkins (link to launch via javaws), deprecating *slave-installer-module and removing these from the WAR, and properly announcing this as a breaking change with a migration guide.

Yeah, I agree that all of these are required for a proper removal. But the very minimum is to remove it from the GUI. All the rest is code cleanup that can and should be done, but breaking the implementation without removing the link from the GUI feels wrong to me, since that is exposing users to broken functionality.

**Comment by timja** [comment]
@basil it’s only shown to users running on Java 8 to my knowledge

**Comment by basil** [comment]
A point already addressed in https://github.com/jenkinsci/remoting/pull/531#issuecomment-1118836708 AFAICT.

**Review by jglick — CHANGES_REQUESTED** [review]
I think we may as well proceed with #532.

**Comment by jglick** [comment]
Up to you. Until it is approved and merged, the discussion on how best to proceed remains open.

**Comment by offa** [comment]
Ok, so I'll leave open here for now. Feel free to close at any time (especially once the other PR is ready / merged).

**Pull request closed without merging**